STA130: HOMEWORK 1 (DUE THURSDAY SEP 12, 2024)

Student name: Sylvia Wong
Student Number: 1010888540
UTORid: wongsyl7

Question 1

In [3]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

# Count the number of missing values in each column
missing_values_count = data.isnull().sum()

# Display the total number of missing values
total_missing = missing_values_count.sum()

print("Missing values in each column:")
print(missing_values_count)
print("\nTotal number of missing values:", total_missing)

Missing values in each column:
row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

Total number of missing values: 12


SUMMARY OF EXCHANGE (QUESTION 1)

Chat History: https://chatgpt.com/share/7228b415-0560-45a2-88bd-0cbee976781e

**Summary of Interaction:**

1. **Request for Code to Import Dataset and Count Missing Values:**
   - You provided a URL to a dataset and asked for Python code to import the data and count missing values.

2. **Provided Solution:**
   - I shared a Python script using the `pandas` library that:
     - Loads the dataset from the provided URL using `pd.read_csv()`.
     - Counts missing values in each column with `data.isnull().sum()`.
     - Sums the total number of missing values across the dataset.
   - The script prints both the count of missing values per column and the total number of missing values.

Question 2 

NOTE: The chat log summary is provided for Question 2, 3, 4, and 5 since these questions are part of the same ChatGPT session.

In [1]:
import pandas as pd

# Load the dataset from the URL
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
data = pd.read_csv(url)

# Determine the number of rows and columns
rows, columns = data.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 391
Number of columns: 11


Definitions

Observations: Observations are individual data entries usually in the form of a row in a data set. These rows contain  information about a specific entity, but what the observations represent may vary depending on the context. For example, an observation could be a row that represents a UofT student and their UTORid, email, and grades.

Variables: Variables are characteristics of a speicific observation. In other words, variables are the columns of the dataset and provide information about the observation. Variables can be quantitative (provides numerical information) or qualitative (provides categorical information). In the above example about a UofT student, the variables would be the student's name (qualitative), UTORid (qualitative), email (qualitative), and grades (quantitative). 

Question 3 

In [3]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

# Display summary statistics of numerical columns
print("Summary of Numerical Columns:")
print(df.describe())

# Display unique value counts for categorical columns
print("\nValue Counts of Categorical Columns:")
for col in df.select_dtypes(include=['object', 'bool']).columns:  # Loop through categorical and boolean columns
    print(f"\nColumn: {col}")
    print(df[col].value_counts())

Summary of Numerical Columns:
            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

Value Counts of Categorical Columns:

Column: id
id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64

Column: name
name
Admiral    1
Muffy      1
Paula      1
Patty      1
Pate       1
          ..
Elvis      1
Eloise     1
Elmer      1
Ellie      1
Zucker     1
Name: count, Length: 391, dtype: int64

Column: gender
gender
male      204
female    187
Name: count, dtype: int64

Column: species
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      1

Question 4 (Using the data set https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv)

Size of data set from df.describe()

In [2]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Display summary statistics of numerical columns
print("Summary of Numerical Columns:")
print(df.describe())

# Display unique value counts for categorical columns including 'alone'
print("\nValue Counts of Categorical Columns:")
for col in df.select_dtypes(include=['object', 'bool']).columns:  # Include categorical and boolean columns
    print(f"\nColumn: {col}")
    print(df[col].value_counts())

Summary of Numerical Columns:
         survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Value Counts of Categorical Columns:

Column: sex
sex
male      577
female    314
Name: count, dtype: int64

Column: embarked
embarked
S    644
C    168
Q     77
Name: count, dtype: int64

Column: class
class
Third     491
First     216
Second    184
Name: count, dty

The Dataset

In [10]:
pd.read_csv(url)

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

All columns

In [4]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

Size of the data set from df.shape()

In [15]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Determine the number of rows and columns
rows, columns = df.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 891
Number of columns: 15


Missing values

In [9]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Count missing values in each column
missing_values = df.isnull().sum()

# Print the number of missing values for each column
print("Missing values in each column:")
print(missing_values)

Missing values in each column:
survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


Explanation: 

a. df.describe() only analyzes numerical columns while df.shape provides the total number of columns in the DataFrame regardless of the value stored. This is because df.describe() must use numerical values to perform calculations (e.g. standard deviation).

In this data set, as seen in the cell that calls "df.columns", there are 15 total columns. However, df.describe() only analyzes 6 columns: survived, pclass, age, sibap, parch, and fare. This is because the remaining 9 columns (sex, embarked, class, who, adult_male, deck, embark_town, alive, alone) are non-numerical columns. In the value counts summary, these columns are analyzed as df.value_counts() analyzes non-numerical data. This explains why df.shape counted 15 columns while df.describe() analyzed only 6 columns. 

b. The discrepancy in the count row comes from the missing values. df.shape finds the size of the Dataframe. This means it counts the total number of rows and columns in the DataFrame regardless of the value stored. However, in the "count" row produced by df.describe(), the number of NON-NULL entries from numerical columns are counted. 

Because, as seen in the cell directly above, the column "age" has 177 missing values, df.describe() ignores these values and thus reports 714 in the "count" for "age" as 714 is 891 (the total number of rows calculated by df.shape) minus 177 (the number of missing values in the age column). Ultimately, the discrepancy between the value in the "count" row of df.describe() and the the total number of rows counted by df.shape is a result of the missing values.

Question 5

The difference between an attribute and a method is that an attribute is a VARIABLE while a method is a FUNCTION. This means that an attribute describes a state of an object while a method defines what actions an object can perform. The reason why attributes do not have the pair of brackets is that attributes are stored data on the object while a method is something the object can do and thus needs to be passed arguments to operate on the object. Ultimately, an attribute only stores information about an object while a method performs a operation on an object to modfiy its attributes. 

SUMMARY OF EXCHANGE (QUESTIONS 2-5)

Chat history: https://chatgpt.com/share/74ad93de-18b6-447f-ae0c-0770ac96ca58

### Summary of Interactions

1. **Dataset Discrepancies in Titanic Dataset**:
   - The user encountered discrepancies in the number of columns when working with different versions of the Titanic dataset. We discussed how different sources might provide slightly altered versions of the dataset with varying numbers of columns.

2. **Counting Rows and Columns**:
   - We explored using `df.shape` to determine the number of rows and columns in a DataFrame, clarifying that this method does not count null values but simply shows the overall dimensions.

3. **Missing Values**:
   - We covered how to count missing values in a dataset using `df.isnull().sum()`, which counts the number of missing entries in each column.

4. **Descriptive Statistics**:
   - The user requested guidance on using `df.describe()` to obtain summary statistics of numerical columns and `df['column'].value_counts()` to summarize categorical data. These functions were explained and demonstrated with the Titanic dataset.

5. **Attributes vs. Methods**:
   - We discussed the difference between attributes (which represent the state or properties of an object) and methods (which represent the behavior or actions an object can perform) in the context of object-oriented programming.

6. **Summarizing the Seaborn Titanic Dataset**:
   - We provided a Python code example to summarize the columns of the Titanic dataset using `df.describe()` for numerical data and `df['column'].value_counts()` for categorical columns, including how to handle boolean columns like "alone."

7. **Summarizing the Animal Crossing Villagers Dataset**:
   - Similar analysis was performed on the Animal Crossing villagers dataset using the same approach to summarize both numerical and categorical columns, highlighting the dataset's structure and key attributes.

8. **Code Examples**:
   - Throughout the exchanges, code examples were provided to load datasets, count rows, handle missing data, summarize numerical and categorical columns, and distinguish between different programming concepts.


Question 6

Count: The count column counts how many non-missing values are in a numerical column

Mean: This value is the average of all the values in the numerical column. That is the sum of all the values divided by the number of values.

Std: This is the standard deviation of the values in the column which measures the amount of variation. 

Min: This displays the minimum value in the column

25%: This displays the value at the 25th percentile or the first quartile meaning 25% of the values in the column are below or at this value.

50%: This displays the median or the value in the middle of the column (when ordered numerically). 50% of the values are less than or equal to this value.

75%; This displays the 75th percentile or the third quartile meaning 75% of the values are less than or equal to this value.

Max: This displays the maximum value in the column.

SUMMARY OF EXCHANGE (QUESTION 6)

Chat history: https://chatgpt.com/share/886871b7-26a7-45b9-8bcb-bc97c10b4d18

### Summary of Interactions

1. **Understanding `df.describe()` Columns:**
   - **count**: Shows the number of non-null (non-missing) entries for each column.
   - **mean**: The average value of the numeric columns.
   - **std**: The standard deviation, measuring the dispersion of values.
   - **min**: The minimum value in the numeric columns.
   - **25%**: The 25th percentile value, where 25% of the data falls below this point.
   - **50%**: The 50th percentile value, or median, where 50% of the data falls below this point.
   - **75%**: The 75th percentile value, where 75% of the data falls below this point.
   - **max**: The maximum value in the numeric columns.
   - For non-numeric columns, `df.describe()` typically includes count, unique (number of unique values), top (most frequent value), and freq (frequency of the most frequent value).

2. **The "count" Column in `df.describe()`:**
   - Displays the number of non-null entries for each column, helping you understand the completeness of the data.

Question 7

In [5]:
#Part 1

import pandas as pd # Example

# Sample survey data with missing values
data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David'],
    'Age': [25, 30, None, 40],
    'Income': [50000, None, 45000, None],
    'Feedback': ['Good', 'Excellent', None, 'Good']
}

df = pd.DataFrame(data)

print("Original DataFrame:")
print(df)

Original DataFrame:
      Name   Age   Income   Feedback
0    Alice  25.0  50000.0       Good
1      Bob  30.0      NaN  Excellent
2  Charlie   NaN  45000.0       None
3    David  40.0      NaN       Good


In [6]:
cleaned_df = df.dropna() #Using df.dropna()

print("\nDataFrame after using dropna():")
print(cleaned_df)


DataFrame after using dropna():
    Name   Age   Income Feedback
0  Alice  25.0  50000.0     Good


In [7]:
del df['Age'] #Using del df['col']
df

Name   Income   Feedback
0    Alice  50000.0       Good
1      Bob      NaN  Excellent
2  Charlie  45000.0       None
3    David      NaN       Good

df.dropna() is used to remove rows with any missing values in it. del df['col'] is used to delete a column from the DataFrame. 

Using df.dropna() is better than using del df['col'] when there are missing values in different columns in the data set. In the example above, using df.dropna() would be preferred because del df['col'] would delete any specified column even if not all values are missing. This would remove important information from rows that do not have missing values. For example, if one performed del df['Age'], the "Age" column would be removed. The value in Age for Charlie is NaN; however, Alice, Bob, and David do not have missing values for Age. Removing the 'Age' column using del df['col'] would therefore remove this important information for Alice, Bob, and David. If this column was important for analysis, it would not be ideal to remove the whole column for all customers. On the other hand df.dropna(), as seen above, retains this information for entries that do not contain missing values (i.e. Alice). Therefore, df.dropna() is more effective in this case.

In [10]:
# Part 2

import pandas as pd

# Sample DataFrame with missing values
data = {
    'CustomerID': [1, 2, 3, 4, 5],
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Eva'],
    'Email': ['alice@example.com', 'bob@example.com', 'charlie@example.com', 'david@example.com', 'eva@example.com'],
    'PhoneNumber': [None, None, '123-456-7890', None, None]
}

df = pd.DataFrame(data)

print("Original DataFrame:")
print(df)

Original DataFrame:
   CustomerID     Name                Email   PhoneNumber
0           1    Alice    alice@example.com          None
1           2      Bob      bob@example.com          None
2           3  Charlie  charlie@example.com  123-456-7890
3           4    David    david@example.com          None
4           5      Eva      eva@example.com          None


In [11]:
# Delete the column with too many missing values
del df['PhoneNumber']

print("\nDataFrame after deleting the 'PhoneNumber' column:")
print(df)


DataFrame after deleting the 'PhoneNumber' column:
   CustomerID     Name                Email
0           1    Alice    alice@example.com
1           2      Bob      bob@example.com
2           3  Charlie  charlie@example.com
3           4    David    david@example.com
4           5      Eva      eva@example.com


You would use del df['col'] over df.dropna() when your data has unwanted columns or column with mostly missing values, rendering the column useless. If you were to use df.dropna(), you would end up losing significant amounts of data as rows with missing values in the useless column would be deleted even if you want to preserve other information for that entry. In other words, if a column is mostly empty, you would use del df['col'] instead of using df.dropna().

In the example above, the column 'PhoneNumber' contains mostly missing values. If you were to use df.dropna() to remove these msising values, you would end up with a table with only the row for "Charlie". This removes crucial information like the emails and names of the other customers and significantly reduces the size of the data sample. So instead, it would be better to use del df['PhoneNumber'] to remove only the "PhoneNumber" column. This not only preserves important information but also streamlines the data as the 'PhoneNumber' column with many missing values would not be useful in analysis anyways.

Ultimately, you would choose df.dropna() or del df['col'] depending on which preserves the most information while removing missing data.

In [12]:
# Part 3

import pandas as pd

# Creating a sample DataFrame with missing values
data = {
    'CustomerID': [1, 2, 3, 4, 5],
    'Name': ['Alice', 'Bob', 'Charlie', 'David', 'Eva'],
    'Email': ['alice@example.com', 'bob@example.com', None, 'david@example.com', 'eva@example.com'],
    'PhoneNumber': [None, None, '123-456-7890', None, None],
    'Notes': [None, 'Called back', None, 'Preferred customer', None]
}

df = pd.DataFrame(data)

print("Original DataFrame:")
print(df)

Original DataFrame:
   CustomerID     Name              Email   PhoneNumber               Notes
0           1    Alice  alice@example.com          None                None
1           2      Bob    bob@example.com          None         Called back
2           3  Charlie               None  123-456-7890                None
3           4    David  david@example.com          None  Preferred customer
4           5      Eva    eva@example.com          None                None


In [13]:
# Delete the 'PhoneNumber' and 'Notes' columns due to high missing values
del df['PhoneNumber']
del df['Notes']

print("\nDataFrame after deleting irrelevant columns:")
print(df)


DataFrame after deleting irrelevant columns:
   CustomerID     Name              Email
0           1    Alice  alice@example.com
1           2      Bob    bob@example.com
2           3  Charlie               None
3           4    David  david@example.com
4           5      Eva    eva@example.com


In [14]:
# Drop rows with any missing values
df_cleaned = df.dropna()

print("\nDataFrame after using dropna():")
print(df_cleaned)


DataFrame after using dropna():
   CustomerID   Name              Email
0           1  Alice  alice@example.com
1           2    Bob    bob@example.com
3           4  David  david@example.com
4           5    Eva    eva@example.com


Part 3:

To remove missing data using both del df['col'] and df.dropna(), you would use del df['col'] to remove unwanted and useless columns with high levels of missing values first. Then you would use df.dropna() to remove any rows that contain missing values. 

It is crucial to use del df['col'] for columns with high proportions of missing data first because, as seen in the example above, using df.dropna() first would remove every single customer from the data set because every entry has at least one missing value in the "useless columns". This significantly decreases the sample size. However, this is not ideal because the name and email of each customer are not missing and should be retained. Instead, using del df['col'] first to remove the columns with a high proportion of missing values (i.e. PhoneNumber and Notes) would retain information about the name and email of the customer and df.dropna() would remove any stragglers (e.g. Charlie) that still have missing values. In other words, del df['col'] removes large amounts of missing values first then df.dropna() removes any remaining missing values.

The key is to use del df['col'] only when the column has a HIGH number of missing values (e.g. more than 50%) and use df.dropna() to remove the remaining entries that still have missing values. This way, information is retained as much as possible.

In [15]:
# Part 4: Before

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [14]:
# Part 4: After (note: I used a for loop because I made my code general to work with any dataset)

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Find Missing Values Per Column
missing_value_per_column = df.isnull().sum()

# Remove Useless Columns
for col in df.columns:
    if (missing_value_per_column[col] / len(df)) > 0.5: # If more than a half of the column is missing, it is removed (the threshold is 50%)
        del (df[col])
    
# Remove Rows with Missing Values
df_cleaned = df.dropna()
print (df_cleaned)

     survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
885         0       3  female  39.0      0      5  29.1250        Q   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male  embark_town alive  alone  


In my approach, I first counted the number of missing values per column. Then, using a for loop, I looped through each column in the dataset and compared the number of missing values in that column to the total number of values in that column. If more than 50% of the values in the column are missing, that column is removed. Then I removed the remaining rows with missing values. 

I used this approach because it minimizes the amount of data removed and retains as much useful information as possible. By checking the proportion of missing values to total data in each column, I can weigh the importance of keeping the column or keeping the entries that contain those missing values. In this case, the "deck" column had more than 50% missing values. Removing the deck column would outweigh removing all the entries with missing values in the deck column because if I were to remove the entries, I would end up removing over 1/2 of the data. This would significantly cut down the size of the data set, giving less to analyze. Although the threshold of "50%" could vary, I chose 50% because I believe that if a column is more than 50% empty, it does not convey too much important information. Furthermore, if you instead used df.dropna(), you would cut down the data set by more than 1/2, resulting in a significantly smaller sample size.

From before, the columns "age", "embarked", "embark_town" and "deck" have missing values. Age, embarked, and embark_town were not removed because the missing values in these column make up less than 50% of the column. This indicates that these column are important enough to keep. Instead, the entries with missing values in the "age", "embark_town", and "embark" column are removed. This would minimize the data removed.

SUMMARY OF INTERACTIONS (Q7)

Chat history: https://chatgpt.com/share/01284cab-bfa8-42b9-a2e9-06e5dbc72732

### **Summary of Interactions**

#### 1. **Understanding `del df['col']` vs. `df.dropna()`**

- **`del df['col']`**:
  - Removes a specific column entirely from the DataFrame.
  - Preferred when you want to remove irrelevant, redundant, or highly missing columns.
  
- **`df.dropna()`**:
  - Removes rows or columns containing missing values based on specified conditions.
  - Use when cleaning the DataFrame by removing rows or columns with missing values without targeting a specific column.

- **Use Case Comparison**:
  - **`df.dropna()`** is better when you want to remove incomplete rows or columns to maintain overall data integrity.
  - **`del df['col']`** is preferred when a specific column has too many missing values or is irrelevant.

#### 2. **Using `del df['col']` and `df.dropna()` Together**

- **Scenario**: First, remove unnecessary columns with many missing values using `del df['col']`, then use `df.dropna()` to further clean rows or columns with remaining missing values.
  
- **Example**:
  ```python
  import pandas as pd
  
  data = {
      'CustomerID': [1, 2, 3, 4, 5],
      'Email': ['alice@example.com', 'bob@example.com', None, 'david@example.com', 'eva@example.com'],
      'PhoneNumber': [None, None, '123-456-7890', None, None],
      'Notes': [None, 'Called back', None, 'Preferred customer', None]
  }
  
  df = pd.DataFrame(data)
  
  # Remove 'PhoneNumber' and 'Notes' columns due to high missing values
  del df['PhoneNumber']
  del df['Notes']
  
  # Remove rows with missing values
  df_cleaned = df.dropna()
  ```

#### 3. **Counting Values in a Column**

- **Methods**:
  - `len(df['col'])`: Counts all values, including missing values.
  - `df['col'].count()`: Counts only non-missing values.
  - `df['col'].value_counts()`: Counts occurrences of each unique value, excluding missing values.

#### 4. **Handling the KeyError: 'col'**

- **Problem**: The error occurs when referencing a column that does not exist.
- **Solution**: Ensure column names used in code are correct and match those in the DataFrame.

- **Example**:
  ```python
  missing_value_per_column = df.isnull().sum()
  
  # Correctly iterate over columns
  for col in df.columns:
      if (missing_value_per_column[col] / len(df)) >= 0.8:
          del df[col]
  ```

#### 5. **Determining When a Column Has Too Many Missing Values**

- **Factors to Consider**:
  - **Threshold**: Commonly used thresholds are 50%, 60-80%, and above 80% missing values.
  - **Importance**: Essential columns may be retained even with missing data.
  - **Data Imputability**: Consider if missing values can be imputed reliably.
  - **Correlation**: Columns highly correlated with others can sometimes be imputed.

- **Example**:
  ```python
  # Calculate the percentage of missing values
  missing_percentage = df.isnull().mean() * 100
  
  # Define a threshold for dropping columns
  threshold = 50
  columns_to_drop = missing_percentage[missing_percentage > threshold].index
  
  # Drop columns exceeding the missing values threshold
  df_cleaned = df.drop(columns=columns_to_drop)
  ```

- **Guidelines**:
  - Use threshold-based criteria to remove columns with excessive missing values.
  - Balance between data retention and the quality of remaining data, focusing on the impact on analysis.

Question 8

Part 1:
The method df.groupby("col1")["col2"].describe() first groups all the values in a specified column ("col1") into categories. For example, the class column is divided into the following groups: "First", "Second" and "Third".

Then this method performs the describe analysis (i.e. count, mean, std, min, max, 25%, 50%, 75%) on "col2" for each group of "col1". For example, if we take the fare column as "col2", the method would calculate the count, mean, std, min, max, 25%, median, and 75% of the fare for First Class passengers, Second Class passengers, and Third Class passengers separately (see below). Ultimately, df.groupby("col1")["col2"].describe groups the data in "col1" and provides the statistics for "col2" within each group of "col1".

In [15]:
import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby('class')['fare'].describe()

count       mean        std  min       25%      50%   75%       max
class                                                                      
First   216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
Second  184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
Third   491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

Part 2:

In conjunction with group_by, count is used to describe how many values of 'col2' are in each category of 'col1'. In the example above, count would determine how much data it has for "Fares" out of the people who are in First Class, out of the people who are in Second Class, and out of the people who are in Third Class. This takes priority over counting the missing values, although missing values are still removed. On the other hand, the count in df.describe() counts how many non-null values there are in each original, numeric column of the entire dataset. In other words, combined with group_by, the values that count analyzes are only the values of 'col2' that are in each group of 'col1', while when used on the whole data set, count analyzes all the values in each column across the whole data frame.

SUMMARY OF EXCHANGE (Question 8 Part 1, 2)

Chat History: https://chatgpt.com/share/fda4e07d-27df-4f30-b6c8-cfcfaa109d63

### **Summary of Interactions on Titanic Dataset Analysis**

1. **Introduction to the Titanic Dataset:**
   - You started working with the Titanic dataset available at `https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv`. This dataset includes information about passengers, such as their age, sex, class, fare, and whether they survived the sinking of the Titanic. The dataset has some missing values, which is typical for real-world data.

2. **Groupby and Describe Explanation:**
   - You were interested in understanding the code `df.groupby("col1")["col2"].describe()`. I explained that this code groups the DataFrame by unique values in `"col1"`, then calculates descriptive statistics for `"col2"` within each group. The descriptive statistics include count, mean, standard deviation, minimum, percentiles (25%, 50%, 75%), and maximum values.

3. **Example with the Titanic Dataset:**
   - To illustrate this concept with the Titanic dataset, we used the code `df.groupby("sex")["age"].describe()`. This groups the data by the `sex` column and calculates descriptive statistics for the `age` column separately for males and females. The output would typically include the number of passengers, average age, and other summary statistics for each gender.

4. **Technical Issue:**
   - An attempt to fetch the data directly during our session failed due to connectivity issues, but the explanation of the code logic and its application remained the same. 

Part 3: Errors

In [1]:
# Missing import statement (A)

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby('class')['fare'].describe()

NameError: name 'pd' is not defined

ChatGPT immediately recognizes that this error is a result of a missing import statement for the Pandas library. It provides the code to fix it which is adding the "import pandas as pd" statement at the top of the code.

On Google, however, it is much more difficult to find the reason for this error. When I searched this error, Google provided StackOverflow for someone else's code and the error that occured with that code. It did not include the explanation that there is a missing import statement. Looking at the second link, you must read through the website to see that you forgot to give pandas the alias of "pd". It is much more difficult to find the explanation that there is a missing import statement.

Thus, it is easier to fix this error in ChatGPT.

In [3]:
# Mistyped titanic.csv 'where only titanics.csv is typed' (B)

import pandas as pd

# Load the dataset
url = "titanics.csv"
df = pd.read_csv(url)

df.groupby('class')['fare'].describe()

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

In [4]:
# Mistyped titanic.csv (where titanic has an 's') (B)

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

df.groupby('class')['fare'].describe()

HTTPError: HTTP Error 404: Not Found

In [1]:
# Mistyped titanic.csv (where 'csv' is missing the 'v') (B)

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.cs"
df = pd.read_csv(url)

df.groupby('class')['fare'].describe()

HTTPError: HTTP Error 404: Not Found

ChatGPT provides the reason for the FileNotFound error and an explanation on how to fix it quickly. It immediately notices that "titanics.csv" is not a file and suggests the correct URL for this data set. It also suggests that you can input the entire path for the file. For the HTTPError, it also suggests the correct link.

A google search is less helpful as it does not automatically suggest the correct URL for the titanics file but it does suggest adding the full path to the url. For the HTTPError, Google does not provide helpful suggestions as you must sort through all the websites and StackOverflow questions to see if any errors are the same as your code.

Thus this error is easier to correct in ChatGPT.

In [5]:
# Misnaming variable (C)

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

DF.groupby('class')['fare'].describe()

NameError: name 'DF' is not defined

ChatGPT explains why there is an error here. It suggests defining the dataframe as DF in order to ensure it is defined before performing any fucntions.

Google also explains why this error occurs. It is slightly more difficult to find the information, especially in StackOverflow, as you must sort through other people's code and examples in order to find the error and an explanation; however, information about this error is easier to find than previous errors.

Ultimately, ChatGPT is more efficient to fix the error; however, in this case, Google Search is still effective.

In [6]:
# Missing parantheses

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url

DF.groupby('class')['fare'].describe()

SyntaxError: '(' was never closed (457507472.py, line 7)

ChatGPT and Google search both explain this error and suggest to fix it by adding a closing paranthesis. However, Google once again brings up StackOverflow and Reddit which are relative to other people's code. The explanations in these cases do not apply to my code. For example, StackOverflow had information about installing a crontab package and a SyntaxError that came from it. This makes Google less effective for fixing this error as it is less tailored to my code and my dataset.

In [9]:
# Typo

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby('class')['fare'].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

ChatGPT immediately notices the typo in the chain function and suggests to use describe() instead. Google does not explain this error and does not realize that there is a typo in the built in function describe(). Once again, Google provides examples of other people's code through StackOverflow making it less useful for my code.

ChatGPT is, therefore, more effective here.

In [11]:
# Non-existent column

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby('Sex')['Age'].describe()

KeyError: 'Sex'

ChatGPT quickly determines that the column name is mistyped or non-existent. It suggests to review the code and look for any typing errors or to use .columns to compare the column "Sex" to the existing columns in the dataset.

Google also identifies that this error is a result of a "key that could not be found" (in the sense of a key-value pair in a dictionary). This makes Google effective; however, it is less focused than ChatGPT as the search also has links about what happens if a KeyError occurs but the column does exist.

ChatGPT is ultimately more effective here.

In [12]:
# Forgetting to put string quotes

import pandas as pd

# Load the dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby(sex)['age'].describe()

NameError: name 'sex' is not defined

For this error, ChatGPT was still able to find the issue for the error as it does suggest the following: "Correct Usage: If you're working with column names directly, you don’t need to define them as variables. Instead, use them as strings in DataFrame operations." However, it does also suggest other possible situations that would result in this error (e.g. Check for Typos: Ensure that 'sex' is correctly spelled and is consistent with the DataFrame column names. For example, if your column name is 'Sex' with a capital 'S', using 'sex' will cause an error.). It also suggests that maybe a variable was not define. In this case, ChatGPT is less focused with its answers and provides a more verbose answer than previously, but it still provides a general list of possibilities as to why the error could occur.

Google, on the other hand, does not give a focused answer. It just suggests how to fix a NameError by defining a variable. StackOverflow also provides other people's code as examples, making the explanations much less useful for my code. Thus it is ineffective for solving this error.

Ultimately, ChatGPT is better than Google Search for trouble shooting as it can provide more focused explanations of the error specific to my code. Google, on the other hand, provides many websites, and I must sort through these websites to find the possible error that occured for my code. Some websites provided (like StackOverflow) are specific to other people's code, making it less useful for my trouble shooting.

SUMMARIES OF EXCHANGE (Question 8)

Chat history: https://chatgpt.com/share/03fcb38e-7895-4b77-ad36-e47041197cd8

### Summary of Interactions

1. **Error: `NameError: name 'pd' is not defined`**
   - **Issue**: You encountered an error because `pd` was not defined.
   - **Solution**: Import the `pandas` library with `import pandas as pd`.

2. **Error: `HTTPError: HTTP Error 404: Not Found`**
   - **Issue**: The URL for the Titanic dataset resulted in a 404 error, meaning the file was not found.
   - **Solution**: Use an alternative URL: `https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv`.

3. **Error: `FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'`**
   - **Issue**: The file `'titanics.csv'` could not be found in the directory.
   - **Solution**: Ensure the file path is correct, verify the file location, or use a URL to download the dataset directly.

4. **Error: `NameError: name 'DF' is not defined`**
   - **Issue**: Attempted to use `DF` without defining it first.
   - **Solution**: Ensure `DF` is assigned with a DataFrame object before use.

5. **Error: `SyntaxError: '(' was never closed`**
   - **Issue**: There was an unclosed parenthesis in the code.
   - **Solution**: Check and correct the syntax by ensuring all parentheses are properly closed.

6. **Error: `AttributeError: 'SeriesGroupBy' object has no attribute 'describle'`**
   - **Issue**: Attempted to use `describle` instead of `describe` on a `SeriesGroupBy` object.
   - **Solution**: Use the correct method `describe()`.

7. **Error: `KeyError: 'col1'`**
   - **Issue**: Column `'col1'` was not found in the DataFrame.
   - **Solution**: Verify column names with `DF.columns` and ensure no typos or case sensitivity issues.

8. **Error: `KeyError: 'Sex'`**
   - **Issue**: Column `'Sex'` was not found in the DataFrame.
   - **Solution**: Verify the column name and ensure it exists in the DataFrame with `DF.columns`.

9. **Error: `NameError: name 'sex' is not defined`**
   - **Issue**: Used `sex` without defining it or without referencing the correct column name.
   - **Solution**: Ensure the correct column name is used and properly defined.
   

Question 9: Yes, the ChatBot helped explain concepts and TA office hours helped to clarify questions.